contracts/parent-chain/contract-upgrades/NitroContracts2Point1Point0UpgradeAction.sol

```diff
@@ -0,0 +1,134 @@
+            "NitroContracts2Point1Point0UpgradeAction: _newRollupAdminLogic is not a contract"
+        );
+        require(
+            Address.isContract(_newRollupUserLogic),
+            "NitroContracts2Point1Point0UpgradeAction: _newRollupUserLogic is not a contract"
+        );
+
+        newWasmModuleRoot = _newWasmModuleRoot;
+        newChallengeManagerImpl = _newChallengeManagerImpl;
+        osp = _osp;
+        condRoot = _condRoot;
+        condOsp = _condOsp;
+        newRollupAdminLogic = _newRollupAdminLogic;
+        newRollupUserLogic = _newRollupUserLogic;
+    }
+
+    function perform(IRollupCore rollup, ProxyAdmin proxyAdmin) external {
+        /// check if previous upgrade v1.2.1 was performed by polling function which was introduced in that version
+        ISequencerInbox_v1_2_1 sequencerInbox = ISequencerInbox_v1_2_1(address(rollup.sequencerInbox()));
+        try sequencerInbox.isUsingFeeToken() returns (bool) {}
+        catch {
+            revert("NitroContracts2Point1Point0UpgradeAction: sequencer inbox needs to be at version >= 1.2.1");
+        }
+
+        /// do the upgrade
+        _upgradeChallengerManager(rollup, proxyAdmin);
+        _upgradeRollup(address(rollup));
+    }
+
+    function _upgradeChallengerManager(IRollupCore rollup, ProxyAdmin proxyAdmin) internal {
+        // set the new challenge manager impl
+        TransparentUpgradeableProxy challengeManager =
+            TransparentUpgradeableProxy(payable(address(rollup.challengeManager())));
+        proxyAdmin.upgradeAndCall(
+            challengeManager,
+            newChallengeManagerImpl,
+            abi.encodeCall(IChallengeManagerUpgradeInit.postUpgradeInit, (osp, condRoot, condOsp))
+        );
+
+        // verify
+        require(
+            proxyAdmin.getProxyImplementation(challengeManager) == newChallengeManagerImpl,
+            "NitroContracts2Point1Point0UpgradeAction: new challenge manager implementation set"
+        );
+        require(
+            IChallengeManagerUpgradeInit(address(challengeManager)).osp() == address(osp),
+            "NitroContracts2Point1Point0UpgradeAction: new OSP not set"
+        );
+
+        // set new wasm module root
+        IRollupAdmin(address(rollup)).setWasmModuleRoot(newWasmModuleRoot);
+
+        // verify:
+        require(
+            rollup.wasmModuleRoot() == newWasmModuleRoot,
+            "NitroContracts2Point1Point0UpgradeAction: wasm module root not set"
+        );
+    }
+
+    function _upgradeRollup(address rollupProxy) internal {
+        IRollupUpgrade rollup = IRollupUpgrade(rollupProxy);
+
+        // set new logic contracts
+        rollup.upgradeTo(newRollupAdminLogic);
+        rollup.upgradeSecondaryTo(newRollupUserLogic);
+
+        // verify
+        require(
+            rollup.anyTrustFastConfirmer() == address(0),
+            "NitroContracts2Point1Point0UpgradeAction: fast confirmer feature not available"
```

Comment:
I would just say unexpected fast confirmer address, as if it is not available it would revert instead

Reply:
https://github.com/OffchainLabs/orbit-actions/pull/16/commits/84e9b7774fb001aa14ed0077b9b5538d2842f479